lib/dentaku/rules.rb

```diff
@@ -35,6 +35,35 @@ def self.each
       @rules.each { |r| yield r }
     end
 
+    def self.refresh
+      @rules ||= core_rules
+      @p_values ||= {}
+      @p_categories ||= {}
+
+      @rules.each do |pattern, _|
+        @p_values[pattern] = pattern.map{ |matcher| matcher.values }.reduce{ |a, e| a.merge(e) }.keys
+        @p_categories[pattern] = pattern.map{ |matcher| matcher.categories }.reduce{ |a, e| a.merge(e) }.keys
+      end
+    end
+
+    def self.filter(tokens)
+      @cache ||= {}
+      categories = tokens.map { |token| token.category }.uniq
+      values = tokens.map { |token| token.value.is_a?(Numeric) ? 0 : token.value }
+      values.select! { |token| ![:fopen, :close].include?(token) }
```

Comment:
Quick note but selecting tokens that are not included in the array seems less obvious (to me) than a positive version:

```ruby
values.reject! { |token| token.in? [:fopen, :close] }
```